### PR TITLE
core: Fix "Failed to get Object" panics

### PR DESCRIPTION
**Pull request by CUB3D**
This constitues the ruffle-side changes of https://github.com/ruffle-rs/rust-flash-lso/pull/67 and https://github.com/ruffle-rs/rust-flash-lso/pull/68. 

This fixes the following (previously would panic on load): 
https://github.com/ruffle-rs/ruffle/issues/17902 - loads
https://github.com/ruffle-rs/ruffle/issues/17709 - loads
https://github.com/ruffle-rs/ruffle/issues/13748 - loads
https://github.com/ruffle-rs/ruffle/issues/10520 - loads

This prevents panics on the following:
https://github.com/ruffle-rs/ruffle/issues/17782 - (and possible dupe https://github.com/ruffle-rs/ruffle/issues/17336 )
Now errors with "Property uploadFromBitmapData not found on flash.display3D.textures.RectangleTexture and there is no default value."

https://github.com/ruffle-rs/ruffle/issues/14329
Fails to load past 54%, some URL load failures might be related

https://github.com/ruffle-rs/ruffle/issues/10174
Same as 17782 above

These are not fixed, but one possible underlying issue is that a new `AMF3Decoder` is created for each call to `ByteArray.readObject` which could result in missing references that are replaced with `Undefined` when an element references a prior value in the stream. In the simple case, we could just cache the `AMF3Decoder` in the `ByteArray` object, but this won't handle more complex cases where the `ByteArray`s position is changed, so I'm leaving this for future work.
Update: I'm actually unsure if the above guess is true anymore, a missing change progressed the above and removed the invalid reference warnings but it still seems odd.

### CODE REVIEW

File: core/Cargo.toml
Comment (torokati44) [inline core/Cargo.toml:46]: This is the head of the currently unmerged https://github.com/ruffle-rs/rust-flash-lso/pull/68, which seems to be unfinished? This seems a bit unfortunate...

Would it be possible to update to the current `master` of that repo first?

File: .gitignore
Comment (torokati44) [inline .gitignore:18]: nit: missing trailing linefeed

File: core/src/avm2/amf.rs
Comment (Lord-McSweeney) [inline core/src/avm2/amf.rs:299]: `deserialize_value_impl` only seems to be called here, can't it be inlined into `deserialize_value`?

File: core/src/avm2/amf.rs
Comment (CUB3D) [reply core/src/avm2/amf.rs]: Changed

File: core/src/avm2/amf.rs
Comment (CUB3D) [reply core/src/avm2/amf.rs:299]: Good catch, I forgot the change the other recursive calls to `deserialize_value_impl`, which has progressed 17782/17336/10174

File: .gitignore
Comment (CUB3D) [reply .gitignore]: Fixed

File: core/Cargo.toml
Comment (CUB3D) [reply core/Cargo.toml:46]: That pr isn't unfinished it's just waiting for review? I don't mind this pr being stalled until that pr is merged, but I'd rather not revert this to a buggier commit

Comment (torokati44) [reply core/Cargo.toml:46]: I see, sorry then. I interpreted "Note this is not as clean as I would like, but I want to fix the panics first and worry about the refactor later" as you still working on it.

**Comment by torokati44** [comment]
Since the act of merging https://github.com/ruffle-rs/rust-flash-lso/pull/68 changed its HEAD hash (and there were some changes added to it even before the merge), I amended this branch accordingly. While at it, I also squashed a few "fixup" type commits into the original commits they fix, for cleaner commit history. I hope you don't mind, @CUB3D.

File: core/src/avm2/amf.rs
Comment (torokati44) [inline core/src/avm2/amf.rs:470]: nit: I believe we generally prefer `expect`, with a meaningful message.

**Review by torokati44 — APPROVED** [review]
Other than that one note, I vote on merging this soon. It looks fairly non-problematic.

**Review by Dinnerbone — APPROVED** [review]
LGTM (but mostly trusting you as the expert here :D)

File: core/src/avm2/amf.rs
Comment (torokati44) [reply core/src/avm2/amf.rs:470]: Actually, I don't think it's worth blocking on this, I see no way for this to fail, `obj` is created locally, right above this line.

Comment (torokati44) [reply core/src/avm2/amf.rs:470]: If you are also okay with leaving this as is, @Dinnerbone, I'd say go for the merge.

File: core/src/avm2/array.rs
Comment (Lord-McSweeney) [inline core/src/avm2/array.rs:130]: Can this panic, just to ensure that it doesn't accidentally get called on a Sparse array?

**Review by Lord-McSweeney — APPROVED** [review]
Approved except for above comment

File: core/src/avm2/array.rs
Comment (CUB3D) [reply core/src/avm2/array.rs]: Done

File: core/src/avm2/amf.rs
Comment (CUB3D) [reply core/src/avm2/amf.rs]: Done

File: core/src/avm2/array.rs
Comment (torokati44) [inline core/src/avm2/array.rs:118]: This comment is now outdated due to the added `panic!`.
I'm just gonna fix this and merge.
Thank you!

Comment (torokati44) [reply core/src/avm2/array.rs]: ... also squashed this last "chore" commit into the first one.

**Pull request merged**
